### PR TITLE
[DEV-6068] Touch update_date for COVID-19 awards not present in latest submissions

### PR DESCRIPTION
**Pull request by collinwr**
**Description:**
This PR adds a management command that updates the `update_date` field on records from the `awards` table if a FABA record is not present in the most recent period's submissions but was present in the period directly before this period

**Technical details:**
A `--all` option will remove the check for covid awards present in last period's submissions and update all covid awards not present in the current period's submissions

**Requirements for PR merge:**

1. [x] Unit & integration tests updated
2. [x] API documentation updated (N/A)
3. [x] Necessary PR reviewers:
    - [x] Backend
4. [x] Matview impact assessment completed (N/A)
5. [x] Frontend impact assessment completed (N/A)
6. [x] Data validation completed
7. [ ] Appropriate Operations ticket(s) created
8. [x] Jira Ticket [DEV-6068](https://federal-spending-transparency.atlassian.net/browse/DEV-6068):
    - [x] Link to this Pull-Request
    - [x] Performance evaluation of affected Script (300s to update 166,017 rows)
    - [x] Before / After data comparison

### CODE REVIEW

**Review by tony-sappe — COMMENTED** [review]
Minor suggestions. The SQL appears to work but could use a little more verification to double-check.

File: usaspending_api/etl/management/commands/update_missing_covid_awards.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+    help = (
+        "This command sets the 'update_date' field on award records with Covid"
+        "faba records present in a submission from the previous submission but"
+        "not in the current period's submission."
+    )
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            "--broad",
```
[the remark this answers is not in the document]
Comment (collinwr) [reply usaspending_api/etl/management/commands/update_missing_covid_awards.py:140]: Updated all usages of "broad" to "all"

File: usaspending_api/etl/management/commands/update_missing_covid_awards.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+WHERE
+    id IN (
+        SELECT
+            award_id
+        FROM
+            covid_awards
+        WHERE
+            covid_awards.is_final_balances_for_fy = FALSE
+    )
+    AND update_date < '{}'
```
[the remark this answers is not in the document]
Comment (collinwr) [reply usaspending_api/etl/management/commands/update_missing_covid_awards.py:109]: Added this in two places

File: usaspending_api/etl/management/commands/update_missing_covid_awards.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+            "last_month": self.read_period_fields(recent_month_periods[1]),
+            "last_quarter": self.read_period_fields(recent_quarter_periods[1]),
+        }
+
+        return recent_periods
+
+    def read_period_fields(self, period):
+        return {"month": period[0], "year": period[1], "submission_reveal_date": period[3]}
+
+    def update_awards(self, periods):
```
[the remark this answers is not in the document]
Comment (collinwr) [reply usaspending_api/etl/management/commands/update_missing_covid_awards.py]: This is done

File: usaspending_api/etl/management/commands/update_missing_covid_awards.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        DISTINCT ON
+        (faba.award_id) faba.award_id,
+        sa.is_final_balances_for_fy
+    FROM
+        financial_accounts_by_awards faba
+    INNER JOIN disaster_emergency_fund_code defc ON
+        defc.code = faba.disaster_emergency_fund_code
+        AND defc.group_name = 'covid_19'
+    INNER JOIN submission_attributes sa ON
+        sa.reporting_period_start >= '2020-04-01'
```
Comment (brett-varney) [inline usaspending_api/etl/management/commands/update_missing_covid_awards.py:20]: Do we need to remember to change this each year?

**Review by brett-varney — COMMENTED** [review]
inline question

File: usaspending_api/etl/management/commands/update_missing_covid_awards.py (hunk 4 above)
Comment (tony-sappe) [reply usaspending_api/etl/management/commands/update_missing_covid_awards.py:20]: No, that is the beginning of valid monthly DABS submission periods

File: usaspending_api/etl/tests/integration/test_touch_old_awards_for_es.py [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+    Award.objects.filter(pk=award_data1).update(update_date=OLD_DATE)
+    original_datetime = Award.objects.get(id=award_data1)
+
+    call_command(SCRIPT_NAME)
+
+    after = Award.objects.get(id=award_data1)
+
+    assert original_datetime.update_date.strftime("%Y-%m-%d") == OLD_DATE, "Sanity check failed"
+    assert original_datetime.update_date != after.update_date, "Award was not properly updated"
+    assert after.update_date - today < timedelta(minutes=1), "New datetime isn't today"
```
Comment (collinwr) [inline usaspending_api/etl/tests/integration/test_touch_old_awards_for_es.py:217]: Could this throw an error if the test is run between 12:00-12:01am?

Comment (tony-sappe) [reply usaspending_api/etl/tests/integration/test_touch_old_awards_for_es.py:217]: Unless I am mistaken, both of these values are datetimes. Can you walk me through the problem?

**Review by kwhickey — COMMENTED** [review]
I'd be okay to approve this given a few things:

1. Because this is such a slippery subject that slips from my mind in just a few weeks, add some details in either comments or help text of the management command around WHEN we run this and HOW. Can be suggestive "need to ..." "Should run this when /as ..."
2. Some dry run stats on what a valid run would produce in prod and qat dbs. If it's not hard, may be good to program in a dry run option that shows how many would be updated

File: usaspending_api/etl/management/commands/update_missing_covid_awards.py [hunk 6]
```diff
@@ -128,13 +130,28 @@
...
 class Command(BaseCommand):
 
     help = (
-        "This command sets the 'update_date' field on award records with Covid"
-        "faba records present in a submission from the previous submission but"
-        "not in the current period's submission."
+        "This command sets the 'update_date' field on award records with Covid "
+        "faba records present in a submission from the previous submission but "
+        "not in the current period's submission. This should be run after each "
+        "period's `submission_reveal_date`. Running the command more frequently"
```
Comment (tony-sappe) [inline usaspending_api/etl/management/commands/update_missing_covid_awards.py:154]: Missing the final line in the comment?

File: usaspending_api/etl/management/commands/update_missing_covid_awards.py [hunk 7]
```diff
@@ -187,9 +220,15 @@ def retreive_recent_periods(self):
...
 
         # Open connection to database
         with connection.cursor() as cursor:
             cursor.execute(update_sql)
-            logger.info(f"Update message (records updated): {cursor.statusmessage}")
+
+            # Log results
+            if count_only:
+                count = cursor.fetchone()[0]
+                logger.info(f"Ready to update awards: {count}")
```
Comment (tony-sappe) [inline usaspending_api/etl/management/commands/update_missing_covid_awards.py:232]:
```suggestion
                logger.info(f"There are {count:,} award records which should be reloaded into Elasticsearch for data consistency.")
```

**Review by tony-sappe — APPROVED** [review]

File: usaspending_api/etl/tests/integration/test_touch_old_awards_for_es.py (hunk 5 above)
Comment (collinwr) [reply usaspending_api/etl/tests/integration/test_touch_old_awards_for_es.py:217]: I looked at this again, and realized I was thinking about it wrong. We're good to go here

**Pull request merged**